src/org/exist/storage/BrokerPools.java

```diff
@@ -0,0 +1,319 @@
+     *
+     * @throws EXistException If the initialization fails.
+     *
+     * @Deprecated Use {@link #configure(String, int, int, Configuration, Optional)}
+     */
+    @Deprecated
+    public static void configure(final String instanceName, final int minBrokers, final int maxBrokers,
+            final Configuration config) throws EXistException {
+        configure(instanceName, minBrokers, maxBrokers, config, Optional.empty());
+    }
+
+    /**
+     * Creates and configures a database instance and adds it to the pool.
+     * Call this before calling {link #getInstance()}.
+     * If a database instance with the same name already exists, the new configuration is ignored.
+     *
+     * @param instanceName A <strong>unique</strong> name for the database instance.
+     *                     It is possible to have more than one database instance (with different configurations
+     *                     for example).
+     * @param minBrokers        The minimum number of concurrent brokers for handling requests on the database instance.
+     * @param maxBrokers        The maximum number of concurrent brokers for handling requests on the database instance.
+     * @param config            The configuration object for the database instance
+     * @param statusObserver    Observes the status of this database instance
+     *
+     * @throws EXistException If the initialization fails.
+     */
+    public static void configure(final String instanceName, final int minBrokers, final int maxBrokers,
+            final Configuration config, final Optional<Observer> statusObserver) throws EXistException {
+
+        // optimize for read-concurrency as instances are configured (created) once and used many times
+        final Lock readLock = instancesLock.readLock();
+        readLock.lock();
+        try {
+            if (instances.containsKey(instanceName)) {
+                LOG.warn("Database instance '" + instanceName + "' is already configured");
+                return;
+            }
+        } finally {
+            readLock.unlock();
+        }
+
+        // fallback to probably having to create a new BrokerPool instance
+        final Lock writeLock = instancesLock.writeLock();
+        writeLock.lock();
+        try {
+            // check again, as another thread may have pre-empted us since we released the read-lock
+            if (instances.containsKey(instanceName)) {
+                LOG.warn("Database instance '" + instanceName + "' is already configured");
+                return;
+            }
+
+            LOG.debug("Configuring database instance '" + instanceName + "'...");
+            try {
+                //Create the instance
+                final BrokerPool instance = new BrokerPool(instanceName, minBrokers, maxBrokers, config, statusObserver);
+                //Add it to the list
+                instances.put(instanceName, instance);
```

Comment:
The problem still is that startup triggers will fail with this PR and we need to solve this somehow. The exceptions are clear: access to the security manager requires BrokerPool.getInstance to return a valid instance. So it's not possible to create users during startup.

I'm not sure how to solve it. Maybe startup triggers should be run from BrokerPools once it has configured a pool instead? Will try this...

```
org.expath.pkg.repo.PackageException: Failed to create user: eXide
        at org.exist.repo.Deployment.checkUserSettings(Deployment.java:620) ~[exist.jar:?]
        at org.exist.repo.Deployment.deploy(Deployment.java:405) ~[exist.jar:?]
        at org.exist.repo.Deployment.installAndDeploy(Deployment.java:271) ~[exist.jar:?]
        at org.exist.repo.AutoDeploymentTrigger.execute(AutoDeploymentTrigger.java:72) [exist.jar:?]
        at org.exist.storage.BrokerPool.callStartupTriggers(BrokerPool.java:774) [exist.jar:?]
        at org.exist.storage.BrokerPool.initialize(BrokerPool.java:686) [exist.jar:?]
        at org.exist.storage.BrokerPool.<init>(BrokerPool.java:398) [exist.jar:?]
        at org.exist.storage.BrokerPools.configure(BrokerPools.java:178) [exist.jar:?]
        at org.exist.storage.BrokerPools.configure(BrokerPools.java:110) [exist.jar:?]
        at org.exist.jetty.JettyStart.run(JettyStart.java:163) [exist-optional.jar:?]
        at org.exist.launcher.Launcher$2.run(Launcher.java:176) [exist.jar:?]
Caused by: org.exist.config.ConfigurationException: org.exist.EXistException: Database instance 'exist' is not available
        at org.exist.security.AbstractPrincipal.<init>(AbstractPrincipal.java:68) ~[exist.jar:?]
        at org.exist.security.AbstractGroup.<init>(AbstractGroup.java:56) ~[exist.jar:?]
        at org.exist.security.internal.GroupImpl.<init>(GroupImpl.java:49) ~[exist.jar:?]
        at org.exist.security.internal.SecurityManagerImpl.addGroup(SecurityManagerImpl.java:561) ~[exist.jar:?]
        at org.exist.repo.Deployment.checkUserSettings(Deployment.java:608) ~[exist.jar:?]
```